sync/src/snapshot/service.rs

```diff
@@ -55,16 +60,88 @@ impl ChainNotify for Service {
             .map(|hash| self.client.block_number(BlockId::Hash(*hash)).expect("Enacted block must exist"))
             .any(|number| number % self.period == 0);
         if is_checkpoint && best_number > self.period {
+            let client = self.client.clone();
             let root_dir = self.root_dir.clone();
             let period = self.period;
             spawn(move || {
-                let target = (best_number / period - 1) * period;
-                let path: PathBuf = [root_dir, target.to_string()].iter().collect();
-                if let Ok(_) = create_dir_all(path) {
-                    // FIXME: implement this
-                    // unimplemented!()
+                let number = (best_number / period - 1) * period;
+                let header = client.block_header(BlockId::Number(number)).expect("Snapshot target must exist");
+                let path: PathBuf = [root_dir, format!("{:x}", header.hash())].iter().collect();
+                match create_dir_all(&path) {
+                    Ok(_) => {}
+                    // FIXME: Handle file system errors
+                    Err(_) => return,
+                }
+                let db = client.database();
+
+                // FIXME: Handle non-existant nodes
+                let root = if let Some(node) = get_node(&db, &header.state_root()) {
```

Comment:
```rust
let state_root = header.state_root();
let node = if let Some(node) = get_node(&db, state_root) {
    node
} else {
```

Reply:
Why do you think that's better?

Reply:
Oh, I see what's the difference. I'll fix this.